site/content/languages/zig.md

```diff
@@ -0,0 +1,213 @@
+export custom `malloc`/`free` pairs to the host.
+
+For instance, the following code exports `malloc`, `free` from Zig's `page_allocator`:
+
+```zig
+const allocator = std.heap.page_allocator;
+
+pub export fn malloc(length: usize) ?[*]u8 {
+    const buff = allocator.alloc(u8, length) catch return null;
+    return buff.ptr;
+}
+
+pub export fn free(buf: [*]u8, length: usize) void {
+    allocator.free(buf[0..length]);
+}
+```
+
+## System Calls
+
+Please read our overview of WebAssembly and
+[System Calls]({{< ref "_index.md#system-calls" >}}). In short, WebAssembly is
+a stack-based virtual machine specification, so operates at a lower level than
+an operating system.
+
+For functionality the operating system would otherwise provide, you must use
+the `wasm32-wasi` target. This imports host functions in
+[WASI]({{< ref "/specs#wasi" >}}).
+
+For example, `zig build-exe -target wasm32-wasi hello.zig` compiles the
+below `main` function into a WASI function exported as `_start`.
+
+```zig
+const std = @import("std");
+const io = std.io;
+const os = std.os;
+const allocator = std.heap.page_allocator;
+const stdout = std.io.getStdOut();
+```
+
+Note: wazero includes an [example WASI project][10] including [source code][11]
+that implements `cat` without any WebAssembly-specific code.
+
+## Concurrency
+
+Please read our overview of WebAssembly and
+[concurrency]({{< ref "_index.md#concurrency" >}}). In short, the current
+WebAssembly specification does not support parallel processing.
+
+## Optimizations
+
+Below are some commonly used configurations that allow optimizing for size or
+performance vs defaults. Note that sometimes one sacrifices the other.
+
+### Binary size
+
+Those with `%.wasm` binary size constraints can change their source, 
+e.g. picking a [different allocator][9b] or set `zig` flags to reduce it.
+
+Source changes:
+* [WebAssembly-only fast allocator][12] is a specialized allocator that 
+trades generality for simplicity, making however specific assumptions 
+on the runtime environment (e.g. single-threaded access to global variables, 
+it direcly calls Wasm-specific intrinsics and it never releases memory)
+
+[`zig` flags][13]:
+Zig provides several flags to control binary size, speed of execution, 
+safety checks. For instance you may use
+* `-ODebug`: Fast build, enabled safety checks, slower runtime performance, 
+  larger binary size
+* `-OReleaseSafe`: Medium runtime performance, enabled safety checks, 
+  slower compilation speed, larger binary size
+* `-OReleaseSmall`: Medium runtime performance, disabled safety checks, 
+  slower compilation speed, smaller binary size
+
+### Performance
+
+Those with runtime performance constraints can change their source or set
+`zig` flags to improve it.
+
+[`zig` flags][13]:
+* `-OReleaseFast`: Enable additional optimizations, possibly at the cost of 
+  increased binary size.
+
+## Frequently Asked Questions
```

Comment:
any other FAQs worth mentioning here?

Reply:
I don't think we've accumulated any questions often enough for them to be frequent, yet :) Once the doc is in, they can be added, IMHO